H2MLauncher.Core/Services/GameDirectoryService.cs

```diff
@@ -0,0 +1,296 @@
+                NotifyFilter = NotifyFilters.CreationTime | NotifyFilters.LastWrite | NotifyFilters.FileName,
+                EnableRaisingEvents = true,
+                IncludeSubdirectories = true
+            };
+            _fileSystemWatcher.Filters.Add("*.ff");
+            _fileSystemWatcher.Filters.Add(CONFIG_MP_FILENAME);
+
+            _fileSystemWatcher.Changed += FileSystemWatcherEvent;
+            _fileSystemWatcher.Created += FileSystemWatcherEvent;
+            _fileSystemWatcher.Deleted += FileSystemWatcherEvent;
+            _fileSystemWatcher.Error += FileSystemWatcher_Error;
+        }
+
+        private void UninitializeWatcher()
+        {
+            if (_fileSystemWatcher is not null)
+            {
+                _fileSystemWatcher.Dispose();
+                _fileSystemWatcher.Changed -= FileSystemWatcherEvent;
+            }
+        }
+
+        private void FileSystemWatcher_Error(object sender, ErrorEventArgs e)
+        {
+            _logger.LogError(e.GetException(), "Error of game directory file system watcher:");
+
+            // reinitialize
+            WatchGameDirectory(_optionsMonitor.CurrentValue);
+        }
+
+        private void FileSystemWatcherEvent(object sender, FileSystemEventArgs e)
+        {
+            try
+            {
+                _logger.LogTrace("Game directory file changed: {path}, {changeType}", e.FullPath, e.ChangeType);
+
+                string currentDirAbsolutePath = Path.GetFullPath(CurrentDir ?? "");
+
+                if (e.FullPath.Equals(Path.Combine(currentDirAbsolutePath, PLAYERS2_DIR, CONFIG_MP_FILENAME)))
+                {
+                    OnConfigFileChanged(e.FullPath);
+                }
+                else if (e.FullPath.StartsWith(Path.Combine(currentDirAbsolutePath, USERMAPS_DIR)))
+                {
+                    OnUsermapsChanged(Path.Combine(currentDirAbsolutePath, USERMAPS_DIR), e.FullPath);
+                }
+                else
+                {
+                    string relativePath = e.FullPath[currentDirAbsolutePath.Length..];
+                    if (relativePath.EndsWith(".ff"))
+                    {
+                        OnFastFileChanged(Path.GetFileName(relativePath), e.FullPath);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error while handling game directory watcher event");
+            }
+        }
+
+        private void OnConfigFileChanged(string path)
+        {
+            if (!File.Exists(path))
+            {
+                CurrentConfigMp = null;
+                ConfigMpChanged?.Invoke(path, null);
+                return;
+            }
+
+            _logger.LogTrace("Config file change detected, parsing...");
+
+            string content;
+
+            // open file with read write share
+            using (var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```

Comment:
Because i want to dispose the file stream right after reading the content, not only at the end

Reply:
Does it not apply that automatically because you don't use the streams anymore afterwards? I thought it would smartly figure that out and essentially, dispose it after your statement regardless.

Reply:
![grafik](https://github.com/user-attachments/assets/2a90ec1b-3b87-45d2-8277-a9eb310d6749)